# **The Battle of Neighbourhoods IBM Capstone Project**

## **Introduction** 

As part of the course, we looked at location data for Toronto and New York, these cities had a lot of replies from the Foursquare API calls, so I decided to look at smaller cities that are expanding and test if a similar comparison can be done with them. The aim is to contribute to future interested parties for example data analysts, Tourists, location data managers, if there are limitations that might need to be resolved using a different API or collecting more data to feed into the requests etc. In this project the comparison is between Toronto and London Ontario as I wanted to have a city which I had to start compiling the data as there was not much location data that was already compiled, but also a city near Toronto so that can also inform anyone wanting to explore further than Toronto example Tourists on other places of interest.

**Toronto** is a multicultural city in the Province of Ontario, it is the most populated city in Canada, it had a population of over 2.7 million in 2016 (Canadian census), also has the highest proportion of minority ethnic population in Canada. Its economy is diversified: Art, Design, Banking & Finance, Life Science, Technology, Aerospace, Fashion, Education, Tourism etc, thus an attractive location for many headquarters of multinational corporations, it welcomes many immigrants annually thus attracting an inflow of talent and opportunities for innovation and expansion as well as having influence on the neighbourhood demographics and tastes. As urban centres grow some people move out to the suburbs to seek other job opportunities or choose to retire elsewhere, others move to be closer to friends and family, others seek a change etc. Where to relocate to depends on many factors for example, job opportunities in their industry, average wages, housing, public/community provisions, crime rates, access to amenities, warmer weather etc.

**London** is a city in Western Ontario which had a population of 383,822 in 2016 according to the Canadian census, it is about 120 miles from Toronto and Detroit. Strong industries are Healthcare with several hospitals located in the city, Education for example the University of Western Ontario (UWO), Insurance and Technology industries. London holds cultural events; Tourists can also explore the settlement changes that happened after the natural disasters and immigration influence over time since the early settlers, for example links to the UK with many similar names and systems. I want to compare Toronto and London Ontario using location data from Foursquare API to look at the different neighbourhoods in these cities, to find the most popular venues in each.

### **Data**

Location data: postcodes, neighbourhood, borough data for Toronto has been taken from a Wikipedia page “List of postal codes of Canada: M - Wikipedia” that was mentioned as part of the course. The data for London Ontario has been taken from another Wikipedia page “List of postal codes of Canada: N - Wikipedia”. 

The Foursquare API will be used to gather data for the different venues to analyse these urban neighbourhoods in Toronto and London Ontario. Longitude and Latitude data for Toronto was provided in week 3 of the course “http://cocl.us/Geospatial_data” and London Ontario geographical coordinates data were gathered from “https://www.geonames.org/”  also geolocator has been used.

Data scrapping and cleaning using python packages. I will cluster the neighbourhoods and choose top 10 most popular venues in each and plot them on a map using Folium to help visualise and interpret the data.

In [10]:
address = 'Toronto, CA'
geolocator = Nominatim(user_agent="CA_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


In [11]:
address = 'London Ontario, CA'
geolocator = Nominatim(user_agent="CA_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of London Ontario are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of London Ontario are 42.9836747, -81.2496068.


In [1]:
import pandas as pd
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import numpy as np

import json

from geopy.geocoders import Nominatim 

import requests 
from pandas.io.json import json_normalize

import matplotlib.cm as cm

import matplotlib.colors as colors

from sklearn.cluster import KMeans

!pip install folium

print("Libraries Imported")

Libraries Imported


In [2]:
import folium

## Gathering the Data

In [3]:
#Using the information provided in week 3 of the course, creating a dataframe for Toronto which will then be merged with the geo data and the final data will be used to send the API calls. 
Toronto = pd.read_html("https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M")
df=Toronto[0]
df.columns
df.head()

Postal Code           Borough              Neighbourhood
0         M1A      Not assigned               Not assigned
1         M2A      Not assigned               Not assigned
2         M3A        North York                  Parkwoods
3         M4A        North York           Victoria Village
4         M5A  Downtown Toronto  Regent Park, Harbourfront

In [4]:
#The data needs to be cleaned to remove missing data, in this case removing "Not assigned" Borough
df=df.drop(index=df[df["Borough"]=="Not assigned"].index)
df.head()

Postal Code           Borough                                Neighbourhood
2         M3A        North York                                    Parkwoods
3         M4A        North York                             Victoria Village
4         M5A  Downtown Toronto                    Regent Park, Harbourfront
5         M6A        North York             Lawrence Manor, Lawrence Heights
6         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

In [5]:
#Postcodes with Longitude and Latitude csv file that was provided, making the Postal code as the index can then later merge the two dataframes on Postal code
postcodes=pd.read_csv("http://cocl.us/Geospatial_data", index_col="Postal Code")
postcodes.head()

Latitude  Longitude
Postal Code                      
M1B          43.806686 -79.194353
M1C          43.784535 -79.160497
M1E          43.763573 -79.188711
M1G          43.770992 -79.216917
M1H          43.773136 -79.239476

In [6]:
#set index on the other dataframe as Postal Code then can merge  
df.set_index("Postal Code")

Borough  \
Postal Code                     
M3A                North York   
M4A                North York   
M5A          Downtown Toronto   
M6A                North York   
M7A          Downtown Toronto   
M9A                 Etobicoke   
M1B               Scarborough   
M3B                North York   
M4B                 East York   
M5B          Downtown Toronto   
M6B                North York   
M9B                 Etobicoke   
M1C               Scarborough   
M3C                North York   
M4C                 East York   
M5C          Downtown Toronto   
M6C                      York   
M9C                 Etobicoke   
M1E               Scarborough   
M4E              East Toronto   
M5E          Downtown Toronto   
M6E                      York   
M1G               Scarborough   
M4G                 East York   
M5G          Downtown Toronto   
M6G          Downtown Toronto   
M1H               Scarborough   
M2H                North York   
M3H                North York   
M4H                 East York   
M5H          Downtown Toronto   
M6H              West Toronto   
M1J               Scarborough   
M2J                North York   
M3J                North York   
M4J                 East York   
M5J          Downtown Toronto   
M6J              West Toronto   
M1K               Scarborough   
M2K                North York   
M3K                North York   
M4K              East Toronto   
M5K          Downtown Toronto   
M6K              West Toronto   
M1L               Scarborough   
M2L                North York   
M3L                North York   
M4L              East Toronto   
M5L          Downtown Toronto   
M6L                North York   
M9L                North York   
M1M               Scarborough   
M2M                North York   
M3M                North York   
M4M              East Toronto   
M5M                North York   
M6M                      York   
M9M                North York   
M1N               Scarborough   
M2N                North York   
M3N                North York   
M4N           Central Toronto   
M5N           Central Toronto   
M6N                      York   
M9N                      York   
M1P               Scarborough   
M2P                North York   
M4P           Central Toronto   
M5P           Central Toronto   
M6P              West Toronto   
M9P                 Etobicoke   
M1R               Scarborough   
M2R                North York   
M4R           Central Toronto   
M5R           Central Toronto   
M6R              West Toronto   
M7R               Mississauga   
M9R                 Etobicoke   
M1S               Scarborough   
M4S           Central Toronto   
M5S          Downtown Toronto   
M6S              West Toronto   
M1T               Scarborough   
M4T           Central Toronto   
M5T          Downtown Toronto   
M1V               Scarborough   
M4V           Central Toronto   
M5V          Downtown Toronto   
M8V                 Etobicoke   
M9V                 Etobicoke   
M1W               Scarborough   
M4W          Downtown Toronto   
M5W          Downtown Toronto   
M8W                 Etobicoke   
M9W                 Etobicoke   
M1X               Scarborough   
M4X          Downtown Toronto   
M5X          Downtown Toronto   
M8X                 Etobicoke   
M4Y          Downtown Toronto   
M7Y              East Toronto   
M8Y                 Etobicoke   
M8Z                 Etobicoke   

                                                 Neighbourhood  
Postal Code                                                     
M3A                                                  Parkwoods  
M4A                                           Victoria Village  
M5A                                  Regent Park, Harbourfront  
M6A                           Lawrence Manor, Lawrence Heights  
M7A                Queen's Park, Ontario Provincial Government  
M9A                    Islington Avenue, Humber Valley Village  
M1B                                 

In [7]:
#merge the dataframes
df_Toronto=pd.merge(df,postcodes,on=["Postal Code"])
df_Toronto.head()

Postal Code           Borough                                Neighbourhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  
0  43.753259 -79.329656  
1  43.725882 -79.315572  
2  43.654260 -79.360636  
3  43.718518 -79.464763  
4  43.662301 -79.389494

In [8]:
#Then plotted the Toronto data using folium, to get a map of Toronto (Fig 1. Map of Toronto)
map_toronto = folium.Map(location=[43.6534817, -79.3839347], zoom_start=11)

# add markers to map
for lat, lng, borough, neighbourhood in zip(df_Toronto['Latitude'], df_Toronto['Longitude'], df_Toronto["Borough"], df_Toronto['Neighbourhood']):
    label = '{}'.format(neighbourhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  

map_toronto

In [9]:
CLIENT_ID = '1P2JFELWESBS3BYQJCQVAUCF0UQY0L3HWW210OB24YVWVWWG' 
CLIENT_SECRET = 'XUUWYEL1BG2ID04JRSEABHJY1VH25ZI3E4VL35Q0YUO1AM5R' 
VERSION = '20200605' 
LIMIT = 500 

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: 1P2JFELWESBS3BYQJCQVAUCF0UQY0L3HWW210OB24YVWVWWG
CLIENT_SECRET:XUUWYEL1BG2ID04JRSEABHJY1VH25ZI3E4VL35Q0YUO1AM5R


Venues near the Toronto neighbourhood geodata are called using “explore” Foursquare API request

In [12]:
#gathering Venues data: Foursquare "explore" call
def getNearbyVenues(names, latitudes, longitudes, radius=5000):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
                   
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
           
        results = requests.get(url).json()["response"]['groups'][0]['items']
         
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighbourhood Latitude', 
                  'Neighbourhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [13]:
#Toronto Venues from API call
Toronto_venues = getNearbyVenues(names=df_Toronto["Neighbourhood"],
                                   latitudes=df_Toronto["Latitude"],
                                   longitudes=df_Toronto["Longitude"]
                                   )

Parkwoods
Victoria Village
Regent Park, Harbourfront
Lawrence Manor, Lawrence Heights
Queen's Park, Ontario Provincial Government
Islington Avenue, Humber Valley Village
Malvern, Rouge
Don Mills
Parkview Hill, Woodbine Gardens
Garden District, Ryerson
Glencairn
West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale
Rouge Hill, Port Union, Highland Creek
Don Mills
Woodbine Heights
St. James Town
Humewood-Cedarvale
Eringate, Bloordale Gardens, Old Burnhamthorpe, Markland Wood
Guildwood, Morningside, West Hill
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Cedarbrae
Hillcrest Village
Bathurst Manor, Wilson Heights, Downsview North
Thorncliffe Park
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Scarborough Village
Fairview, Henry Farm, Oriole
Northwood Park, York University
East Toronto, Broadview North (Old East York)
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
Kennedy Park, Ionview, East Birchmo

In [14]:
print(Toronto_venues.shape)
Toronto_venues.head()

(10258, 7)


Neighbourhood  Neighbourhood Latitude  Neighbourhood Longitude  \
0     Parkwoods               43.753259               -79.329656   
1     Parkwoods               43.753259               -79.329656   
2     Parkwoods               43.753259               -79.329656   
3     Parkwoods               43.753259               -79.329656   
4     Parkwoods               43.753259               -79.329656   

                         Venue  Venue Latitude  Venue Longitude  \
0              Allwyn's Bakery       43.759840       -79.324719   
1  Donalda Golf & Country Club       43.752816       -79.342741   
2         Galleria Supermarket       43.753520       -79.349518   
3                  Tim Hortons       43.760668       -79.326368   
4                 Island Foods       43.745866       -79.346035   

         Venue Category  
0  Caribbean Restaurant  
1           Golf Course  
2           Supermarket  
3                  Café  
4  Caribbean Restaurant

**The API call returned 10,258 venues in Toronto, a neighbourhood count is done as a snapshot to see if the data is spread across the neighbourhoods.**


In [15]:
Toronto_venues.groupby('Neighbourhood').count()

Neighbourhood Latitude  \
Neighbourhood                                                                
Agincourt                                                              100   
Alderwood, Long Branch                                                 100   
Bathurst Manor, Wilson Heights, Downsview North                        100   
Bayview Village                                                         95   
Bedford Park, Lawrence Manor East                                      100   
Berczy Park                                                            100   
Birch Cliff, Cliffside West                                            100   
Brockton, Parkdale Village, Exhibition Place                           100   
Business reply mail Processing Centre, South Ce...                     100   
CN Tower, King and Spadina, Railway Lands, Harb...                     100   
Caledonia-Fairbanks                                                    100   
Canada Post Gateway Processing Centre                                  100   
Cedarbrae                                                              100   
Central Bay Street                                                     100   
Christie                                                               100   
Church and Wellesley                                                   100   
Clarks Corners, Tam O'Shanter, Sullivan                                100   
Cliffside, Cliffcrest, Scarborough Village West                        100   
Commerce Court, Victoria Hotel                                         100   
Davisville                                                             100   
Davisville North                                                       100   
Del Ray, Mount Dennis, Keelsdale and Silverthorn                       100   
Don Mills                                                              200   
Dorset Park, Wexford Heights, Scarborough Town ...                     100   
Downsview                                                              400   
Dufferin, Dovercourt Village                                           100   
East Toronto, Broadview North (Old East York)                          100   
Eringate, Bloordale Gardens, Old Burnhamthorpe,...                     100   
Fairview, Henry Farm, Oriole                                           100   
First Canadian Place, Underground city                                 100   
Forest Hill North & West, Forest Hill Road Park                        100   
Garden District, Ryerson                                               100   
Glencairn                                                              100   
Golden Mile, Clairlea, Oakridge                                        100   
Guildwood, Morningside, West Hill                                      100   
Harbourfront East, Union Station, Toronto Islands                      100   
High Park, The Junction South                                          100   
Hillcrest Village                                                      100   
Humber Summit                                                          100   
Humberlea, Emery                                                        93   
Humewood-Cedarvale                                                     100   
India Bazaar, The Beaches West                                         100   
Islington Avenue, Humber Valley Village                                100   
Kennedy Park, Ionview, East Birchmount Park                            100   
Kensington Market, Chinatown, Grange Park                              100   
Kingsview Village, St. Phillips, Martin Grove G...                     100   
Lawrence Manor, Lawrence Heights                                       100   
Lawrence Park                                                          100   
Leaside                                                                100   
Little Portugal, Trinity                                               100

In [16]:
print('There are {} uniques categories.'.format(len(Toronto_venues['Venue Category'].unique())))

There are 255 uniques categories.


In [17]:
#The venues are assigned against each neighbourhood so that can later be quantified to work out the popular venues
Toronto_onehot = pd.get_dummies(Toronto_venues[['Venue Category']], prefix="", prefix_sep="")

Toronto_onehot['Neighbourhood'] = Toronto_venues['Neighbourhood'] 

fixed_columns = [Toronto_onehot.columns[-1]] + list(Toronto_onehot.columns[:-1])
Toronto_onehot = Toronto_onehot[fixed_columns]

Toronto_onehot.head()

Neighbourhood  Afghan Restaurant  African Restaurant  Airport  \
0     Parkwoods                  0                   0        0   
1     Parkwoods                  0                   0        0   
2     Parkwoods                  0                   0        0   
3     Parkwoods                  0                   0        0   
4     Parkwoods                  0                   0        0   

   Airport Lounge  American Restaurant  Amphitheater  Art Gallery  \
0               0                    0             0            0   
1               0                    0             0            0   
2               0                    0             0            0   
3               0                    0             0            0   
4               0                    0             0            0   

   Arts & Crafts Store  Asian Restaurant  Athletics & Sports  Auto Dealership  \
0                    0                 0                   0                0   
1                    0                 0                   0                0   
2                    0                 0                   0                0   
3                    0                 0                   0                0   
4                    0                 0                   0                0   

   Automotive Shop  BBQ Joint  Baby Store  Badminton Court  Bagel Shop  \
0                0          0           0                0           0   
1                0          0           0                0           0   
2                0          0           0                0           0   
3                0          0           0                0           0   
4                0          0           0                0           0   

   Bakery  Bank  Bar  Baseball Field  Baseball Stadium  Basketball Court  \
0       0     0    0               0                 0                 0   
1       0     0    0               0                 0                 0   
2       0     0    0               0                 0                 0   
3       0     0    0               0                 0                 0   
4       0     0    0               0                 0                 0   

   Bath House  Beach  Beer Bar  Beer Store  Big Box Store  Bike Shop  Bistro  \
0           0      0         0           0              0          0       0   
1           0      0         0           0              0          0       0   
2           0      0         0           0              0          0       0   
3           0      0         0           0              0          0       0   
4           0      0         0           0              0          0       0   

   Bookstore  Botanical Garden  Boutique  Bowling Alley  Brazilian Restaurant  \
0          0                 0         0              0                     0   
1          0                 0         0              0                     0   
2          0                 0         0              0                     0   
3          0                 0         0              0                     0   
4          0                 0         0              0                     0   

   Breakfast Spot  Brewery  Bridal Shop  Bridge  Bubble Tea Shop  Buffet  \
0               0        0            0       0                0       0   
1               0        0            0       0                0       0   
2               0        0            0       0                0       0   
3               0        0            0       0                0       0   
4               0        0            0       0                0       0   

   Burger Joint  Burrito Place  Butcher  Café  Camera Store  Campground  \
0             0              0        0     0             0           0   
1             0              0        0     0             0           0   
2             0              0        0     0             0           0   
3             0              0        0     1             0           

In [18]:
#As noted 10,258 venues were returned across the Toronto neighbourhoods
Toronto_onehot.shape

(10258, 256)

**The mean value per venue is calculated to rank the popularity and to summarise the data, then a snapshot of the top 5 venues per neighbourhood is viewed.**

In [19]:
#The mean value per venue is calculated to rank the popularity and to summarise the data, then a snapshot of the top 5 venues per neighbourhood is viewed
Toronto_grouped = Toronto_onehot.groupby('Neighbourhood').mean().reset_index()
Toronto_grouped

Neighbourhood  Afghan Restaurant  \
0                                           Agincourt              0.000   
1                              Alderwood, Long Branch              0.000   
2     Bathurst Manor, Wilson Heights, Downsview North              0.000   
3                                     Bayview Village              0.000   
4                   Bedford Park, Lawrence Manor East              0.000   
5                                         Berczy Park              0.000   
6                         Birch Cliff, Cliffside West              0.000   
7        Brockton, Parkdale Village, Exhibition Place              0.000   
8   Business reply mail Processing Centre, South C...              0.000   
9   CN Tower, King and Spadina, Railway Lands, Har...              0.000   
10                                Caledonia-Fairbanks              0.000   
11              Canada Post Gateway Processing Centre              0.010   
12                                          Cedarbrae              0.000   
13                                 Central Bay Street              0.000   
14                                           Christie              0.000   
15                               Church and Wellesley              0.000   
16            Clarks Corners, Tam O'Shanter, Sullivan              0.000   
17    Cliffside, Cliffcrest, Scarborough Village West              0.000   
18                     Commerce Court, Victoria Hotel              0.000   
19                                         Davisville              0.010   
20                                   Davisville North              0.000   
21   Del Ray, Mount Dennis, Keelsdale and Silverthorn              0.000   
22                                          Don Mills              0.005   
23  Dorset Park, Wexford Heights, Scarborough Town...              0.000   
24                                          Downsview              0.000   
25                       Dufferin, Dovercourt Village              0.000   
26      East Toronto, Broadview North (Old East York)              0.000   
27  Eringate, Bloordale Gardens, Old Burnhamthorpe...              0.000   
28                       Fairview, Henry Farm, Oriole              0.000   
29             First Canadian Place, Underground city              0.000   
30    Forest Hill North & West, Forest Hill Road Park              0.000   
31                           Garden District, Ryerson              0.000   
32                                          Glencairn              0.000   
33                    Golden Mile, Clairlea, Oakridge              0.000   
34                  Guildwood, Morningside, West Hill              0.000   
35  Harbourfront East, Union Station, Toronto Islands              0.000   
36                      High Park, The Junction South              0.000   
37                                  Hillcrest Village              0.000   
38                                      Humber Summit              0.000   
39                                   Humberlea, Emery              0.000   
40                                 Humewood-Cedarvale              0.000   
41                     India Bazaar, The Beaches West              0.000   
42            Islington Avenue, Humber Valley Village              0.000   
43        Kennedy Park, Ionview, East Birchmount Park              0.000   
44          Kensington Market, Chinatown, Grange Park              0.000   
45  Kingsview Village, St. Phillips, Martin Grove ...              0.000   
46                   Lawrence Manor, Lawrence Heights              0.000   
47                                      Lawrence Park              0.000   
48                                            Leaside              0.000   
49                           Little Portugal, Trinity              0.000   
50                                     Malvern, Rouge              0.000   
51  Milliken, Agincourt North, Steeles East, L'Amo...              0.000   
52  Mimico N

In [20]:
#top 5 venues
num_top_venues = 5

for hood in Toronto_grouped['Neighbourhood']:
    print("----"+hood+"----")
    temp = Toronto_grouped[Toronto_grouped['Neighbourhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Agincourt----
                  venue  freq
0           Coffee Shop  0.05
1    Chinese Restaurant  0.05
2  Caribbean Restaurant  0.04
3          Noodle House  0.04
4      Sushi Restaurant  0.03


----Alderwood, Long Branch----
           venue  freq
0         Bakery  0.06
1    Coffee Shop  0.06
2    Pizza Place  0.03
3  Burrito Place  0.03
4   Burger Joint  0.03


----Bathurst Manor, Wilson Heights, Downsview North----
                       venue  freq
0                Coffee Shop  0.06
1  Middle Eastern Restaurant  0.04
2          Korean Restaurant  0.03
3              Grocery Store  0.03
4            Bubble Tea Shop  0.03


----Bayview Village----
                       venue  freq
0                Coffee Shop  0.08
1                Supermarket  0.07
2                     Bakery  0.05
3  Middle Eastern Restaurant  0.05
4          Korean Restaurant  0.04


----Bedford Park, Lawrence Manor East----
              venue  freq
0       Coffee Shop  0.09
1              Café  0.07
2    

                venue  freq
0                Café  0.10
1                Park  0.10
2         Coffee Shop  0.09
3  Italian Restaurant  0.08
4       Grocery Store  0.05


----Rouge Hill, Port Union, Highland Creek----
            venue  freq
0     Zoo Exhibit  0.18
1            Park  0.08
2     Coffee Shop  0.06
3     Gas Station  0.04
4  Breakfast Spot  0.03


----Runnymede, Swansea----
         venue  freq
0         Café  0.10
1       Bakery  0.08
2         Park  0.07
3  Coffee Shop  0.07
4      Brewery  0.06


----Runnymede, The Junction North----
                venue  freq
0                Café  0.10
1  Italian Restaurant  0.07
2              Bakery  0.06
3             Brewery  0.06
4         Coffee Shop  0.05


----Scarborough Village----
              venue  freq
0       Coffee Shop  0.08
1      Burger Joint  0.05
2              Park  0.05
3    Ice Cream Shop  0.04
4  Department Store  0.03


----South Steeles, Silverstone, Humbergate, Jamestown, Mount Olive, Beaumond Heights, Th

In [21]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

**The top 10 venues are then returned per neighbourhood, the geo data is added back to the data to be able to plot the neighbourhoods.**

In [22]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

neighbourhoods_venues_sorted = pd.DataFrame(columns=columns)
neighbourhoods_venues_sorted['Neighbourhood'] = Toronto_grouped['Neighbourhood']

for ind in np.arange(Toronto_grouped.shape[0]):
    neighbourhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(Toronto_grouped.iloc[ind, :], num_top_venues)

neighbourhoods_venues_sorted.head()

Neighbourhood 1st Most Common Venue  \
0                                        Agincourt           Coffee Shop   
1                           Alderwood, Long Branch                Bakery   
2  Bathurst Manor, Wilson Heights, Downsview North           Coffee Shop   
3                                  Bayview Village           Coffee Shop   
4                Bedford Park, Lawrence Manor East           Coffee Shop   

       2nd Most Common Venue      3rd Most Common Venue 4th Most Common Venue  \
0         Chinese Restaurant       Caribbean Restaurant          Noodle House   
1                Coffee Shop                Pizza Place                  Park   
2  Middle Eastern Restaurant             Clothing Store     Korean Restaurant   
3                Supermarket  Middle Eastern Restaurant                Bakery   
4                       Café                       Park         Grocery Store   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0                  Park           Supermarket      Sushi Restaurant   
1          Liquor Store    Seafood Restaurant         Grocery Store   
2       Bubble Tea Shop         Grocery Store            Restaurant   
3     Korean Restaurant   Japanese Restaurant       Thai Restaurant   
4      Sushi Restaurant                Bakery   Japanese Restaurant   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0        Sandwich Place                Bakery            Pizza Place  
1          Burger Joint         Burrito Place                Brewery  
2      Sushi Restaurant                Bakery    Japanese Restaurant  
3            Restaurant            Bagel Shop     Chinese Restaurant  
4          Liquor Store         Shopping Mall        Thai Restaurant

In [23]:
kclusters = 6

Toronto_grouped_clustering = Toronto_grouped.drop('Neighbourhood', 1)

kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(Toronto_grouped_clustering)

kmeans.labels_[0:10] 

array([3, 3, 3, 3, 1, 2, 1, 4, 1, 2], dtype=int32)

**The data was grouped into 6 clusters as it gave a better distribution, then the clusters are plotted with folium on the Toronto map.**

In [24]:
#venue data and geodata is merged
neighbourhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)
Toronto_merged = df_Toronto

Toronto_merged = Toronto_merged.join(neighbourhoods_venues_sorted.set_index('Neighbourhood'), on='Neighbourhood')

Toronto_merged.head()

Postal Code           Borough                                Neighbourhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  Cluster Labels      1st Most Common Venue  \
0  43.753259 -79.329656               3  Middle Eastern Restaurant   
1  43.725882 -79.315572               3                       Park   
2  43.654260 -79.360636               2                Coffee Shop   
3  43.718518 -79.464763               1                Coffee Shop   
4  43.662301 -79.389494               2                Coffee Shop   

  2nd Most Common Venue  3rd Most Common Venue 4th Most Common Venue  \
0         Grocery Store   Caribbean Restaurant          Burger Joint   
1           Supermarket           Burger Joint           Coffee Shop   
2                  Park                   Café           Pizza Place   
3    Italian Restaurant  Vietnamese Restaurant                  Café   
4                  Park                   Café        Farmers Market   

  5th Most Common Venue 6th Most Common Venue      7th Most Common Venue  \
0           Supermarket   Japanese Restaurant         Italian Restaurant   
1    Italian Restaurant         Grocery Store  Middle Eastern Restaurant   
2                 Plaza     French Restaurant                      Hotel   
3                Bakery                  Park                       Bank   
4                Bakery                 Plaza             Sandwich Place   

      8th Most Common Venue 9th Most Common Venue    10th Most Common Venue  
0                    Bakery          Liquor Store                      Café  
1              Liquor Store           Pizza Place                Restaurant  
2              Neighborhood        Farmers Market  Mediterranean Restaurant  
3  Mediterranean Restaurant  Brazilian Restaurant              Burger Joint  
4              Neighborhood            Restaurant              Liquor Store

**Toronto Cluster Map below show most clusters fit per geographical area**

In [25]:
#map of Toronto clusters
map_cluster = folium.Map(location=[43.6532, -79.3832], zoom_start=11)

x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

markers_colors = []
for lat, lon, poi, cluster in zip(Toronto_merged['Latitude'], Toronto_merged['Longitude'], Toronto_merged['Neighbourhood'], Toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_cluster)
map_cluster

In [27]:
#Red Cluster
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 0, Toronto_merged.columns[[2] + list(range(6, Toronto_merged.shape[1]))]]

Neighbourhood 1st Most Common Venue  \
6                           Malvern, Rouge           Zoo Exhibit   
12  Rouge Hill, Port Union, Highland Creek           Zoo Exhibit   
95                             Upper Rouge           Zoo Exhibit   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
6            Coffee Shop              Pharmacy           Gas Station   
12                  Park           Coffee Shop           Gas Station   
95        Sandwich Place                  Bank  Fast Food Restaurant   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
6         Sandwich Place  Fast Food Restaurant           Pizza Place   
12                  Bank              Pharmacy        Breakfast Spot   
95           Coffee Shop              Pharmacy         Grocery Store   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
6    Fried Chicken Joint            Restaurant           Burger Joint  
12           Pizza Place         Grocery Store                    Pub  
95           Pizza Place           Gas Station           Liquor Store

In [31]:
#Purple Cluster
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 1, Toronto_merged.columns[[2] + list(range(6, Toronto_merged.shape[1]))]]

Neighbourhood 1st Most Common Venue  \
3                     Lawrence Manor, Lawrence Heights           Coffee Shop   
8                      Parkview Hill, Woodbine Gardens                  Park   
14                                    Woodbine Heights           Coffee Shop   
19                                         The Beaches                  Park   
35       East Toronto, Broadview North (Old East York)                  Park   
41                        The Danforth West, Riverdale                  Park   
44                     Golden Mile, Clairlea, Oakridge                  Park   
47                      India Bazaar, The Beaches West                  Park   
54                                     Studio District           Coffee Shop   
55                   Bedford Park, Lawrence Manor East           Coffee Shop   
56    Del Ray, Mount Dennis, Keelsdale and Silverthorn                  Café   
58                         Birch Cliff, Cliffside West                 Beach   
61                                       Lawrence Park           Coffee Shop   
63                       Runnymede, The Junction North                  Café   
81                                  Runnymede, Swansea                  Café   
88        New Toronto, Mimico South, Humber Bay Shores                  Park   
98       The Kingsway, Montgomery Road, Old Mill North           Coffee Shop   
100  Business reply mail Processing Centre, South C...                  Park   
101  Old Mill South, King's Mill Park, Sunnylea, Hu...    Italian Restaurant   
102  Mimico NW, The Queensway West, South of Bloor,...                  Park   

    2nd Most Common Venue  3rd Most Common Venue 4th Most Common Venue  \
3      Italian Restaurant  Vietnamese Restaurant                  Café   
8                   Beach                   Café                Bakery   
14                Brewery                   Café                  Park   
19            Coffee Shop                  Beach                Bakery   
35            Coffee Shop                Brewery                  Café   
41            Coffee Shop                   Café               Brewery   
44                  Beach            Coffee Shop         Grocery Store   
47            Coffee Shop                Brewery                  Café   
54                   Park                Brewery                  Café   
55                   Café                   Park         Grocery Store   
56     Italian Restaurant                Brewery                Bakery   
58                   Park                 Bakery        Ice Cream Shop   
61                   Park     Italian Restaurant                  Café   
63     Italian Restaurant                 Bakery               Brewery   
81                 Bakery                   Park           Coffee Shop   
88            Coffee Shop     Italian Restaurant      Sushi Restaurant   
98                Brewery                 Bakery                  Park   
100           Coffee Shop                Brewery                  Café   
101                Bakery                   Park                  Café   
102                Bakery            Coffee Shop    Seafood Restaurant   

    5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
3                  Bakery                  Park                  Bank   
8               BBQ Joint               Brewery         Grocery Store   
14              Gastropub             BBQ Joint                Bakery   
19                Brewery        Ice Cream Shop                  Café   
35                 Bakery             BBQ Joint                 Beach   
41       Greek Restaurant                Bakery        Farmers Market   
44         Breakfast Spot        Sandwich Place          Burger Joint   
47                  Beach                Bakery             BBQ Joint   
54         Farmers Market                 Beach                Bakery   
55       Sushi Restaurant                Bakery   Japanese Restaurant   
56

In [32]:
#Blue Cluster, all are in Downtown Toronto
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 2, Toronto_merged.columns[[2] + list(range(6, Toronto_merged.shape[1]))]]

Neighbourhood 1st Most Common Venue  \
2                           Regent Park, Harbourfront           Coffee Shop   
4         Queen's Park, Ontario Provincial Government           Coffee Shop   
9                            Garden District, Ryerson           Coffee Shop   
15                                     St. James Town           Coffee Shop   
20                                        Berczy Park           Coffee Shop   
24                                 Central Bay Street           Coffee Shop   
30                           Richmond, Adelaide, King           Coffee Shop   
36  Harbourfront East, Union Station, Toronto Islands           Coffee Shop   
42           Toronto Dominion Centre, Design Exchange           Coffee Shop   
48                     Commerce Court, Victoria Hotel           Coffee Shop   
80                     University of Toronto, Harbord           Coffee Shop   
84          Kensington Market, Chinatown, Grange Park           Coffee Shop   
87  CN Tower, King and Spadina, Railway Lands, Har...           Coffee Shop   
91                                           Rosedale           Coffee Shop   
92                                     Stn A PO Boxes           Coffee Shop   
96                        St. James Town, Cabbagetown           Coffee Shop   
97             First Canadian Place, Underground city           Coffee Shop   
99                               Church and Wellesley           Coffee Shop   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
2                   Park                  Café           Pizza Place   
4                   Park                  Café        Farmers Market   
9                   Park                  Café        Farmers Market   
15                  Park                  Café        Farmers Market   
20                  Park                  Café                Bakery   
24                  Park                  Café                Bakery   
30                  Park                  Café                Bakery   
36                  Park                  Café                Bakery   
42                  Park                  Café        Farmers Market   
48                  Park                  Café        Farmers Market   
80                  Café                  Park        Sandwich Place   
84                  Park                  Café                Bakery   
87                  Park                  Café                Bakery   
91                  Café                  Park    Italian Restaurant   
92                  Park                 Plaza        Farmers Market   
96                  Park                  Café        Farmers Market   
97                  Park                  Café        Farmers Market   
99                  Café                  Park        Farmers Market   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
2                  Plaza     French Restaurant                 Hotel   
4                 Bakery                 Plaza        Sandwich Place   
9                  Plaza                 Hotel            Restaurant   
15                 Plaza                Bakery                 Hotel   
20        Farmers Market                 Plaza                 Hotel   
24        Farmers Market                 Plaza        Sandwich Place   
30        Farmers Market        Sandwich Place                 Plaza   
36                 Plaza        Farmers Market         Historic Site   
42                Bakery                 Plaza            Restaurant   
48                Bakery                 Plaza    Mexican Restaurant   
80                Bakery           Pizza Place                   Spa   
84        Farmers Market          Liquor Store        Sandwich Place   
87        Farmers Market                 Beach            Restaurant   
91                   Spa        Farmers Market         Historic Site   
92                  Café                Bakery                 Hotel   
96         Histor

In [33]:
#Turquoise Cluster
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 3, Toronto_merged.columns[[2] + list(range(6, Toronto_merged.shape[1]))]]

Neighbourhood  \
0                                           Parkwoods   
1                                    Victoria Village   
7                                           Don Mills   
11  West Deane Park, Princess Gardens, Martin Grov...   
13                                          Don Mills   
17  Eringate, Bloordale Gardens, Old Burnhamthorpe...   
27                                  Hillcrest Village   
28    Bathurst Manor, Wilson Heights, Downsview North   
32                                Scarborough Village   
33                       Fairview, Henry Farm, Oriole   
34                    Northwood Park, York University   
38        Kennedy Park, Ionview, East Birchmount Park   
39                                    Bayview Village   
40                                          Downsview   
45                           York Mills, Silver Hills   
46                                          Downsview   
49           North Park, Maple Leaf Park, Upwood Park   
52                            Willowdale, Newtonbrook   
53                                          Downsview   
59                        Willowdale, Willowdale East   
60                                          Downsview   
64                                             Weston   
65  Dorset Park, Wexford Heights, Scarborough Town...   
66                                    York Mills West   
71                                  Wexford, Maryvale   
72                        Willowdale, Willowdale West   
76              Canada Post Gateway Processing Centre   
78                                          Agincourt   
82            Clarks Corners, Tam O'Shanter, Sullivan   
85  Milliken, Agincourt North, Steeles East, L'Amo...   
90                      Steeles West, L'Amoreaux West   
93                             Alderwood, Long Branch   

        1st Most Common Venue      2nd Most Common Venue  \
0   Middle Eastern Restaurant              Grocery Store   
1                        Park                Supermarket   
7                        Park  Middle Eastern Restaurant   
11                Coffee Shop                     Bakery   
13                       Park  Middle Eastern Restaurant   
17              Grocery Store                Coffee Shop   
27         Chinese Restaurant                     Bakery   
28                Coffee Shop  Middle Eastern Restaurant   
32                Coffee Shop               Burger Joint   
33         Chinese Restaurant                     Bakery   
34                Coffee Shop  Middle Eastern Restaurant   
38                Coffee Shop                       Park   
39                Coffee Shop                Supermarket   
40                Coffee Shop             Clothing Store   
45        Japanese Restaurant                       Café   
46                Coffee Shop             Clothing Store   
49                Coffee Shop             Clothing Store   
52                Coffee Shop          Korean Restaurant   
53                Coffee Shop             Clothing Store   
59          Korean Restaurant                       Café   
60                Coffee Shop             Clothing Store   
64                Coffee Shop     Furniture / Home Store   
65       Caribbean Restaurant  Middle Eastern Restaurant   
66                Coffee Shop           Sushi Restaurant   
71  Middle Eastern Restaurant       Caribbean Restaurant   
72                Coffee Shop  Middle Eastern Restaurant   
76                      Hotel        Japanese Restaurant   
78                Coffee Shop         Chinese Restaurant   
82       Caribbean Restaurant                     Bakery   
85         Chinese Restaurant                     Bakery   
90         Chinese Restaurant                     Bakery   
93                     Bakery                Coffee Shop   

        3rd Most Common Venue      4th Most Common Venue  \
0        Caribbean Restaurant               Burger Joint   
1                Burger Joint                Coffee Shop 

In [34]:
#Green Cluster
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 4, Toronto_merged.columns[[2] + list(range(6, Toronto_merged.shape[1]))]]

Neighbourhood 1st Most Common Venue  \
10                                          Glencairn                  Café   
16                                 Humewood-Cedarvale                  Café   
21                                Caledonia-Fairbanks                  Café   
23                                            Leaside                  Park   
25                                           Christie                  Café   
29                                   Thorncliffe Park                  Park   
31                       Dufferin, Dovercourt Village                  Café   
37                           Little Portugal, Trinity                  Park   
43       Brockton, Parkdale Village, Exhibition Place                  Park   
62                                           Roselawn                  Café   
67                                   Davisville North           Coffee Shop   
68    Forest Hill North & West, Forest Hill Road Park                  Café   
69                      High Park, The Junction South                  Café   
73                  North Toronto West, Lawrence Park                  Park   
74                The Annex, North Midtown, Yorkville                  Café   
75                             Parkdale, Roncesvalles                  Café   
79                                         Davisville                  Park   
83                        Moore Park, Summerhill East                  Park   
86  Summerhill West, Rathnelly, South Hill, Forest...                  Café   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
10           Coffee Shop                  Park               Brewery   
16           Coffee Shop                  Park    Italian Restaurant   
21           Coffee Shop               Brewery                  Park   
23                  Café    Italian Restaurant           Coffee Shop   
25           Coffee Shop                  Park                Bakery   
29                  Café    Italian Restaurant           Coffee Shop   
31           Coffee Shop                   Bar                  Park   
37                  Café           Coffee Shop                Bakery   
43                  Café           Coffee Shop                Bakery   
62                  Park           Coffee Shop    Italian Restaurant   
67                  Park                  Café         Grocery Store   
68           Coffee Shop                  Park         Grocery Store   
69                Bakery           Coffee Shop    Italian Restaurant   
73           Coffee Shop                  Café    Italian Restaurant   
74           Coffee Shop                  Park         Grocery Store   
75                  Park                Bakery           Coffee Shop   
79    Italian Restaurant                  Café           Coffee Shop   
83                  Café           Coffee Shop    Italian Restaurant   
86                  Park           Coffee Shop    Italian Restaurant   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
10    Italian Restaurant                Bakery          Liquor Store   
16                   Bar         Grocery Store                Bakery   
21    Italian Restaurant                   Bar                Bakery   
23                Bakery         Grocery Store             Bookstore   
25    Italian Restaurant         Grocery Store        Sandwich Place   
29                Bakery         Grocery Store             BBQ Joint   
31    Italian Restaurant               Brewery        Ice Cream Shop   
37        Sandwich Place        Farmers Market             Gastropub   
43        Sandwich Place                   Bar           Pizza Place   
62         Grocery Store               Brewery     Indian Restaurant   
67    Italian Restaurant             BBQ Joint           Supermarket   
68                Bakery    Italian Restaurant               Brewery   
69               Brewery                  Park                   Bar   
73        

In [35]:
#Orange Cluster 
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 5, Toronto_merged.columns[[2] + list(range(6, Toronto_merged.shape[1]))]]

Neighbourhood 1st Most Common Venue  \
5             Islington Avenue, Humber Valley Village           Coffee Shop   
18                  Guildwood, Morningside, West Hill           Coffee Shop   
22                                             Woburn           Coffee Shop   
26                                          Cedarbrae           Coffee Shop   
50                                      Humber Summit           Coffee Shop   
51    Cliffside, Cliffcrest, Scarborough Village West           Coffee Shop   
57                                   Humberlea, Emery           Coffee Shop   
70                                          Westmount           Coffee Shop   
77  Kingsview Village, St. Phillips, Martin Grove ...           Coffee Shop   
89  South Steeles, Silverstone, Humbergate, Jamest...           Coffee Shop   
94                Northwest, West Humber - Clairville           Coffee Shop   

   2nd Most Common Venue  3rd Most Common Venue 4th Most Common Venue  \
5         Sandwich Place                Brewery         Grocery Store   
18                  Bank         Sandwich Place                  Park   
22              Pharmacy                   Park           Pizza Place   
26        Sandwich Place            Pizza Place  Fast Food Restaurant   
50                 Hotel         Sandwich Place              Pharmacy   
51                  Park          Grocery Store        Sandwich Place   
57           Pizza Place  Vietnamese Restaurant           Gas Station   
70        Sandwich Place                   Bank           Pizza Place   
77                 Hotel             Restaurant              Pharmacy   
89  Fast Food Restaurant         Sandwich Place     Indian Restaurant   
94                 Hotel             Restaurant    Chinese Restaurant   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
5                 Bakery              Pharmacy           Pizza Place   
18              Pharmacy            Restaurant     Indian Restaurant   
22  Fast Food Restaurant                 Hotel        Sandwich Place   
26                  Bank            Restaurant              Pharmacy   
50     Indian Restaurant            Steakhouse    Italian Restaurant   
51              Pharmacy                   Gym           Pizza Place   
57  Fast Food Restaurant                  Bank              Pharmacy   
70                Bakery            Restaurant           Golf Course   
77        Sandwich Place                Bakery                  Bank   
89    Italian Restaurant  Caribbean Restaurant    Chinese Restaurant   
94  Fast Food Restaurant   Rental Car Location            Steakhouse   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
5           Liquor Store    Italian Restaurant             Restaurant  
18  Fast Food Restaurant   Fried Chicken Joint         Ice Cream Shop  
22            Restaurant         Grocery Store         Breakfast Spot  
26   Fried Chicken Joint  Caribbean Restaurant         Ice Cream Shop  
50    Chinese Restaurant          Burger Joint                   Bank  
51   Japanese Restaurant        Clothing Store                   Bank  
57         Grocery Store    Chinese Restaurant       Asian Restaurant  
70           Gas Station    Chinese Restaurant                  Hotel  
77    Chinese Restaurant           Golf Course     Italian Restaurant  
89              Pharmacy      Asian Restaurant       Sushi Restaurant  
94          Burger Joint     Indian Restaurant         Airport Lounge

**London Ontario** data was then collected from GeoNames for co-ordinates and neighbourhood data,  List of postal codes of Canada: N - Wikipedia for postcodes and neighbourhood data and checking against City Map Gallery (london.ca). 
London Ontario data was not provided by the course and a search on the internet did not result in a tidy table that could be used to read the data and create the dataframe.

In [36]:
#London Ontario dataframe creating by using data collected from geonames and wikipedia, these weren't provided by the course and a search on the internet didnt result in a tidy table that 
#could be used to read the data and create the dataframe, hence this was made below.
LondonOntario={
            "Postal Code":["N6B","N5W","N6A","N6C", "N6H","N6E","N6P", "N6N", "N6M", "N6L", "N6K", "N6J", "N6G", "N5V", "N5Z", "N5Y", "N5X"],
            "Borough":["London Central","London East", "London North", "London South", "London West", "London", "London","London", "London", "London", "London", "London", "London", "London", "London", "London", "London"], 
            "Neighbourhood":["London Central","London East", "London North", "London South", "London West","South White Oaks,Central Westminster,East Longwoods,West Brockley","Talbot,Lambeth,West Tempo,South Sharon Creek","South Highbury,Glanworth,East Brockley,SE Westminster","Jackson,Old Victoria,Bradley,North Highbury","East Tempo","Riverbend,Woodhull,North Sharon Creek,Byron,West Westmount","Southcrest,East Westmount,West Highland","Sunningdale,West Masonville,Medway,NE Hyde Park,East Fox Hollow","YXU,North and East Argyle,East Huron Heights","Glen Cairn","West Huron Heights,Carling","Fanshawe,Stoneybrook,Stoney Creek,Uplands,East Masonville"],
            "Latitude":["42.984","42.986","42.998","42.958","42.991","42.918","42.891","42.9","42.963","42.872","42.954","42.955","43.015","43.023","42.966","43.012","43.044"],
            "Longitude":["-81.239","-81.182","-81.256","-81.238","-81.34","-81.224","-81.324","-81.159","-81.139","-81.247","-81.342","-81.273","-81.305","-81.164","-81.205","-81.231","-81.239"]
}

df_W=pd.DataFrame(LondonOntario)
df_W

Postal Code         Borough  \
0          N6B  London Central   
1          N5W     London East   
2          N6A    London North   
3          N6C    London South   
4          N6H     London West   
5          N6E          London   
6          N6P          London   
7          N6N          London   
8          N6M          London   
9          N6L          London   
10         N6K          London   
11         N6J          London   
12         N6G          London   
13         N5V          London   
14         N5Z          London   
15         N5Y          London   
16         N5X          London   

                                        Neighbourhood Latitude Longitude  
0                                      London Central   42.984   -81.239  
1                                         London East   42.986   -81.182  
2                                        London North   42.998   -81.256  
3                                        London South   42.958   -81.238  
4                                         London West   42.991    -81.34  
5   South White Oaks,Central Westminster,East Long...   42.918   -81.224  
6        Talbot,Lambeth,West Tempo,South Sharon Creek   42.891   -81.324  
7   South Highbury,Glanworth,East Brockley,SE West...     42.9   -81.159  
8         Jackson,Old Victoria,Bradley,North Highbury   42.963   -81.139  
9                                          East Tempo   42.872   -81.247  
10  Riverbend,Woodhull,North Sharon Creek,Byron,We...   42.954   -81.342  
11            Southcrest,East Westmount,West Highland   42.955   -81.273  
12  Sunningdale,West Masonville,Medway,NE Hyde Par...   43.015   -81.305  
13       YXU,North and East Argyle,East Huron Heights   43.023   -81.164  
14                                         Glen Cairn   42.966   -81.205  
15                         West Huron Heights,Carling   43.012   -81.231  
16  Fanshawe,Stoneybrook,Stoney Creek,Uplands,East...   43.044   -81.239

**The London map plotted with the data above shows the locations plotted are scattered across the city and this is expected as only one location data was plotted per neighbourhood.**

In [38]:
map_london = folium.Map(location=[42.9836747, -81.2496068], zoom_start=12)

# add markers to map
for lat, lng, borough, neighbourhood in zip(df_W['Latitude'], df_W['Longitude'], df_W["Borough"], df_W['Neighbourhood']):
    label = '{}'.format(neighbourhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_london)  

map_london

In [39]:
#London venues using the “explore" Foursquare API are then sorted and grouped like was done with Toronto data
London_venues = getNearbyVenues(names=df_W["Neighbourhood"],
                                   latitudes=df_W["Latitude"],
                                   longitudes=df_W["Longitude"]
                                 )

London Central
London East
London North
London South
London West
South White Oaks,Central Westminster,East Longwoods,West Brockley
Talbot,Lambeth,West Tempo,South Sharon Creek
South Highbury,Glanworth,East Brockley,SE Westminster
Jackson,Old Victoria,Bradley,North Highbury
East Tempo
Riverbend,Woodhull,North Sharon Creek,Byron,West Westmount
Southcrest,East Westmount,West Highland
Sunningdale,West Masonville,Medway,NE Hyde Park,East Fox Hollow
YXU,North and East Argyle,East Huron Heights
Glen Cairn
West Huron Heights,Carling
Fanshawe,Stoneybrook,Stoney Creek,Uplands,East Masonville


In [40]:
print(London_venues.shape)
London_venues.head()

(1214, 7)


Neighbourhood Neighbourhood Latitude Neighbourhood Longitude  \
0  London Central                 42.984                 -81.239   
1  London Central                 42.984                 -81.239   
2  London Central                 42.984                 -81.239   
3  London Central                 42.984                 -81.239   
4  London Central                 42.984                 -81.239   

                             Venue  Venue Latitude  Venue Longitude  \
0               Edgar + Joe's Cafe       42.979983       -81.243563   
1              The Morrissey House       42.985865       -81.241693   
2      TG's Addis Ababa Restaurant       42.987210       -81.237187   
3  The Works Gourmet Burger Bistro       42.982151       -81.249710   
4               Abruzzi Ristorante       42.982105       -81.250735   

       Venue Category  
0                Café  
1                 Pub  
2  African Restaurant  
3          Restaurant  
4  Italian Restaurant

**Only 1,214 London venues were returned, this is 11% of the Toronto API replies and this is not surprising as the population in London was about 14% of that of Toronto in 2016.
Also the geo data used to send the request was not the same size as London is a smaller city, this also could explain why some neighbourhoods have only a few venues returned. Another reason could be more geodata needs to be added or the location data needs to be updated**

In [41]:
London_venues.groupby('Neighbourhood').count()

Neighbourhood Latitude  \
Neighbourhood                                                                
East Tempo                                                               4   
Fanshawe,Stoneybrook,Stoney Creek,Uplands,East ...                      81   
Glen Cairn                                                             100   
Jackson,Old Victoria,Bradley,North Highbury                             25   
London Central                                                         100   
London East                                                            100   
London North                                                           100   
London South                                                           100   
London West                                                             90   
Riverbend,Woodhull,North Sharon Creek,Byron,Wes...                      56   
Southcrest,East Westmount,West Highland                                100   
South Highbury,Glanworth,East Brockley,SE Westm...                      13   
South White Oaks,Central Westminster,East Longw...                      96   
Sunningdale,West Masonville,Medway,NE Hyde Park...                     100   
Talbot,Lambeth,West Tempo,South Sharon Creek                            10   
West Huron Heights,Carling                                             100   
YXU,North and East Argyle,East Huron Heights                            39   

                                                    Neighbourhood Longitude  \
Neighbourhood                                                                 
East Tempo                                                                4   
Fanshawe,Stoneybrook,Stoney Creek,Uplands,East ...                       81   
Glen Cairn                                                              100   
Jackson,Old Victoria,Bradley,North Highbury                              25   
London Central                                                          100   
London East                                                             100   
London North                                                            100   
London South                                                            100   
London West                                                              90   
Riverbend,Woodhull,North Sharon Creek,Byron,Wes...                       56   
Southcrest,East Westmount,West Highland                                 100   
South Highbury,Glanworth,East Brockley,SE Westm...                       13   
South White Oaks,Central Westminster,East Longw...                       96   
Sunningdale,West Masonville,Medway,NE Hyde Park...                      100   
Talbot,Lambeth,West Tempo,South Sharon Creek                             10   
West Huron Heights,Carling                                              100   
YXU,North and East Argyle,East Huron Heights                             39   

                                                    Venue  Venue Latitude  \
Neighbourhood                                                               
East Tempo                                              4               4   
Fanshawe,Stoneybrook,Stoney Creek,Uplands,East ...     81              81   
Glen Cairn                                            100             100   
Jackson,Old Victoria,Bradley,North Highbury            25              25   
London Central                                        100             100   
London East                                           100             100   
London North                                          100             100   
London South                                          100             100   
London West                                            90              90   
Riverbend,Woodhull,North Sharon Creek,Byron,Wes...     56              56   
Southcrest,East Westmount,West Highland               100             100   
South Highbury,Glanworth,East Brockley,SE Westm...     13         

In [42]:
print('There are {} uniques categories.'.format(len(London_venues['Venue Category'].unique())))

There are 129 uniques categories.


**Sorting the data:** venues are then displayed by the type and then grouped by the mean so that the most common venue can be grouped together like was done with the Toronto data.

In [43]:
London_onehot = pd.get_dummies(London_venues[['Venue Category']], prefix="", prefix_sep="")

London_onehot['Neighbourhood'] = London_venues['Neighbourhood'] 

fixed_columns = [London_onehot.columns[-1]] + list(London_onehot.columns[:-1])
London_onehot = London_onehot[fixed_columns]

London_onehot.head()

Neighbourhood  African Restaurant  American Restaurant  Arcade  \
0  London Central                   0                    0       0   
1  London Central                   0                    0       0   
2  London Central                   1                    0       0   
3  London Central                   0                    0       0   
4  London Central                   0                    0       0   

   Arts & Crafts Store  Asian Restaurant  Athletics & Sports  BBQ Joint  \
0                    0                 0                   0          0   
1                    0                 0                   0          0   
2                    0                 0                   0          0   
3                    0                 0                   0          0   
4                    0                 0                   0          0   

   Bakery  Bank  Bar  Baseball Field  Baseball Stadium  Beer Store  Bistro  \
0       0     0    0               0                 0           0       0   
1       0     0    0               0                 0           0       0   
2       0     0    0               0                 0           0       0   
3       0     0    0               0                 0           0       0   
4       0     0    0               0                 0           0       0   

   Bookstore  Bowling Alley  Breakfast Spot  Brewery  Bubble Tea Shop  \
0          0              0               0        0                0   
1          0              0               0        0                0   
2          0              0               0        0                0   
3          0              0               0        0                0   
4          0              0               0        0                0   

   Burger Joint  Burrito Place  Bus Station  Café  Camera Store  Campground  \
0             0              0            0     1             0           0   
1             0              0            0     0             0           0   
2             0              0            0     0             0           0   
3             0              0            0     0             0           0   
4             0              0            0     0             0           0   

   Chinese Restaurant  Chocolate Shop  Church  Clothing Store  Coffee Shop  \
0                   0               0       0               0            0   
1                   0               0       0               0            0   
2                   0               0       0               0            0   
3                   0               0       0               0            0   
4                   0               0       0               0            0   

   Convenience Store  Cosmetics Shop  Deli / Bodega  Department Store  \
0                  0               0              0                 0   
1                  0               0              0                 0   
2                  0               0              0                 0   
3                  0               0              0                 0   
4                  0               0              0                 0   

   Dessert Shop  Diner  Discount Store  Dog Run  Eastern European Restaurant  \
0             0      0               0        0                            0   
1             0      0               0        0                            0   
2             0      0               0        0                            0   
3             0      0               0        0                            0   
4             0      0               0        0                            0   

   Electronics Store  Event Space  Farmers Market  Fast Food Restaurant  \
0                  0            0               0                     0   
1                  0            0               0                     0   
2                  0            0               0                     0   
3                  0            0               0           

In [44]:
London_grouped = London_onehot.groupby('Neighbourhood').mean().reset_index()
London_grouped

Neighbourhood  African Restaurant  \
0                                          East Tempo                0.00   
1   Fanshawe,Stoneybrook,Stoney Creek,Uplands,East...                0.00   
2                                          Glen Cairn                0.01   
3         Jackson,Old Victoria,Bradley,North Highbury                0.00   
4                                      London Central                0.01   
5                                         London East                0.01   
6                                        London North                0.01   
7                                        London South                0.01   
8                                         London West                0.00   
9   Riverbend,Woodhull,North Sharon Creek,Byron,We...                0.00   
10            Southcrest,East Westmount,West Highland                0.01   
11  South Highbury,Glanworth,East Brockley,SE West...                0.00   
12  South White Oaks,Central Westminster,East Long...                0.00   
13  Sunningdale,West Masonville,Medway,NE Hyde Par...                0.00   
14       Talbot,Lambeth,West Tempo,South Sharon Creek                0.00   
15                         West Huron Heights,Carling                0.01   
16       YXU,North and East Argyle,East Huron Heights                0.00   

    American Restaurant    Arcade  Arts & Crafts Store  Asian Restaurant  \
0              0.000000  0.000000             0.000000          0.000000   
1              0.012346  0.012346             0.012346          0.024691   
2              0.020000  0.000000             0.000000          0.010000   
3              0.040000  0.000000             0.000000          0.000000   
4              0.010000  0.000000             0.000000          0.010000   
5              0.020000  0.000000             0.000000          0.010000   
6              0.000000  0.000000             0.000000          0.010000   
7              0.010000  0.000000             0.000000          0.010000   
8              0.011111  0.000000             0.000000          0.000000   
9              0.017857  0.000000             0.000000          0.000000   
10             0.010000  0.000000             0.000000          0.010000   
11             0.000000  0.000000             0.000000          0.000000   
12             0.041667  0.000000             0.010417          0.000000   
13             0.010000  0.010000             0.010000          0.010000   
14             0.000000  0.000000             0.000000          0.000000   
15             0.000000  0.000000             0.000000          0.020000   
16             0.000000  0.000000             0.000000          0.025641   

    Athletics & Sports  BBQ Joint    Bakery      Bank   Bar  Baseball Field  \
0                 0.00   0.000000  0.000000  0.000000  0.00        0.000000   
1                 0.00   0.000000  0.000000  0.024691  0.00        0.012346   
2                 0.01   0.000000  0.000000  0.000000  0.02        0.000000   
3                 0.00   0.000000  0.000000  0.040000  0.00        0.000000   
4                 0.01   0.000000  0.010000  0.000000  0.01        0.000000   
5                 0.01   0.010000  0.020000  0.030000  0.01        0.000000   
6                 0.01   0.000000  0.010000  0.000000  0.01        0.010000   
7                 0.01   0.000000  0.000000  0.000000  0.00        0.000000   
8                 0.00   0.011111  0.011111  0.022222  0.00        0.000000   
9                 0.00   0.000000  0.000000  0.035714  0.00        0.000000   
10                0.00   0.000000  0.000000  0.010000  0.01        0.000000   
11                0.00   0.000000  0.000000  0.000000  0.00        0.000000   
12                0.00   0.000000  0.000000  0.010417  0.00        0.000000   
13                0.00   0.010000  0.010000  0.020000  0.00        0.010000   
14                0.00   0.000000  0.000000  0.100000  0.00        0.000000   
15                0.

In [45]:
num_top_venues = 5

for hood in London_grouped['Neighbourhood']:
    print("----"+hood+"----")
    temp = London_grouped[London_grouped['Neighbourhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----East Tempo----
                venue  freq
0               Hotel  0.25
1  Italian Restaurant  0.25
2         Golf Course  0.25
3      Breakfast Spot  0.25
4           Irish Pub  0.00


----Fanshawe,Stoneybrook,Stoney Creek,Uplands,East Masonville----
           venue  freq
0     Restaurant  0.09
1    Coffee Shop  0.07
2            Pub  0.05
3  Grocery Store  0.05
4       Pharmacy  0.04


----Glen Cairn----
           venue  freq
0    Coffee Shop  0.11
1     Restaurant  0.04
2  Grocery Store  0.04
3   Liquor Store  0.04
4           Café  0.03


----Jackson,Old Victoria,Bradley,North Highbury----
         venue  freq
0  Coffee Shop  0.16
1     Pharmacy  0.08
2  Gas Station  0.08
3  Supermarket  0.08
4        Diner  0.04


----London Central----
         venue  freq
0  Coffee Shop  0.06
1          Pub  0.05
2         Café  0.04
3    Gastropub  0.03
4      Brewery  0.03


----London East----
                  venue  freq
0           Coffee Shop  0.14
1  Fast Food Restaurant  0.06
2    

In [51]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

neighbourhoods_venues_sorted = pd.DataFrame(columns=columns)
neighbourhoods_venues_sorted['Neighbourhood'] = London_grouped['Neighbourhood']

for ind in np.arange(London_grouped.shape[0]):
    neighbourhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(London_grouped.iloc[ind, :], num_top_venues)

neighbourhoods_venues_sorted.head()

Neighbourhood 1st Most Common Venue  \
0                                         East Tempo                 Hotel   
1  Fanshawe,Stoneybrook,Stoney Creek,Uplands,East...            Restaurant   
2                                         Glen Cairn           Coffee Shop   
3        Jackson,Old Victoria,Bradley,North Highbury           Coffee Shop   
4                                     London Central           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0    Italian Restaurant           Golf Course        Breakfast Spot   
1           Coffee Shop                   Pub         Grocery Store   
2         Grocery Store          Liquor Store            Restaurant   
3           Supermarket           Gas Station              Pharmacy   
4                   Pub                  Café          Liquor Store   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0                 Diner        Discount Store               Dog Run   
1              Pharmacy        Breakfast Spot     Electronics Store   
2                   Gym           Pizza Place                 Diner   
3         Grocery Store          Intersection           Flower Shop   
4                  Park               Brewery             Gastropub   

         8th Most Common Venue          9th Most Common Venue  \
0  Eastern European Restaurant                   Dessert Shop   
1                         Bank                 Ice Cream Shop   
2           Italian Restaurant                        Brewery   
3         Fast Food Restaurant                  Movie Theater   
4                  Pizza Place  Vegetarian / Vegan Restaurant   

  10th Most Common Venue  
0      Electronics Store  
1           Liquor Store  
2                   Café  
3                Dog Run  
4     Italian Restaurant

In [60]:
kclusters = 6

London_grouped_clustering = London_grouped.drop('Neighbourhood', 1)

kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(London_grouped_clustering)

kmeans.labels_[0:10] 

array([2, 1, 4, 3, 4, 1, 4, 4, 1, 1], dtype=int32)

In [56]:
#London Venues data merged with geodata and clustered
London_merged = df_W

London_merged = London_merged.join(neighbourhoods_venues_sorted.set_index('Neighbourhood'), on='Neighbourhood')

London_merged.head()

Postal Code         Borough   Neighbourhood Latitude Longitude  \
0         N6B  London Central  London Central   42.984   -81.239   
1         N5W     London East     London East   42.986   -81.182   
2         N6A    London North    London North   42.998   -81.256   
3         N6C    London South    London South   42.958   -81.238   
4         N6H     London West     London West   42.991    -81.34   

   Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0               1           Coffee Shop                   Pub   
1               1           Coffee Shop  Fast Food Restaurant   
2               1                   Pub           Coffee Shop   
3               1           Coffee Shop            Restaurant   
4               1           Coffee Shop         Grocery Store   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0                  Café          Liquor Store                  Park   
1           Gas Station              Pharmacy         Grocery Store   
2                  Park                  Café    Italian Restaurant   
3                  Café                  Park    Italian Restaurant   
4                  Park  Fast Food Restaurant           Pizza Place   

           6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0                        Brewery             Gastropub           Pizza Place   
1                            Gym            Restaurant          Liquor Store   
2  Vegetarian / Vegan Restaurant           Pizza Place         Grocery Store   
3                    Pizza Place          Liquor Store         Grocery Store   
4                    Golf Course                   Gym              Pharmacy   

           9th Most Common Venue     10th Most Common Venue  
0  Vegetarian / Vegan Restaurant         Italian Restaurant  
1                           Bank             Sandwich Place  
2                     Restaurant          Indian Restaurant  
3                            Pub  Middle Eastern Restaurant  
4                    Supermarket             Sandwich Place

In [61]:
#London Ontario Cluster Map
map_clusters = folium.Map(location=[42.9836747, -81.2496068 ], zoom_start=11)

x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

markers_colors = []
for lat, lon, poi, cluster in zip(London_merged['Latitude'], London_merged['Longitude'], London_merged['Neighbourhood'], London_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=6,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
map_clusters 

**London Ontario Cluster Map** the clusters also fit per geographical area: the purple cluster is centrally located to the north where the Education and Healthcare industry is based, also the Airport. Three other smaller clusters are in the south of the city

In [63]:
#Red Cluster
London_merged.loc[London_merged['Cluster Labels'] == 0, London_merged.columns[[2] + list(range(6, London_merged.shape[1]))]]

Neighbourhood 1st Most Common Venue  \
7  South Highbury,Glanworth,East Brockley,SE West...           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
7        Breakfast Spot        Sandwich Place        Waste Facility   

  5th Most Common Venue 6th Most Common Venue   7th Most Common Venue  \
7       Warehouse Store        Clothing Store  Thrift / Vintage Store   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
7       Motorcycle Shop  Fast Food Restaurant            Gas Station

In [64]:
#Purple Cluster, this is most of the city 
London_merged.loc[London_merged['Cluster Labels'] == 1, London_merged.columns[[2] + list(range(6, London_merged.shape[1]))]]

Neighbourhood 1st Most Common Venue  \
0                                      London Central           Coffee Shop   
1                                         London East           Coffee Shop   
2                                        London North                   Pub   
3                                        London South           Coffee Shop   
4                                         London West           Coffee Shop   
5   South White Oaks,Central Westminster,East Long...           Coffee Shop   
10  Riverbend,Woodhull,North Sharon Creek,Byron,We...           Coffee Shop   
11            Southcrest,East Westmount,West Highland                  Park   
12  Sunningdale,West Masonville,Medway,NE Hyde Par...           Coffee Shop   
13       YXU,North and East Argyle,East Huron Heights           Coffee Shop   
14                                         Glen Cairn           Coffee Shop   
15                         West Huron Heights,Carling                   Pub   
16  Fanshawe,Stoneybrook,Stoney Creek,Uplands,East...            Restaurant   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                    Pub                  Café          Liquor Store   
1   Fast Food Restaurant           Gas Station              Pharmacy   
2            Coffee Shop                  Park                  Café   
3             Restaurant                  Café                  Park   
4          Grocery Store                  Park  Fast Food Restaurant   
5             Restaurant   Sporting Goods Shop   American Restaurant   
10           Golf Course         Grocery Store                  Park   
11           Coffee Shop           Pizza Place    Italian Restaurant   
12         Grocery Store            Restaurant                   Pub   
13            Restaurant                 Diner                 Hotel   
14         Grocery Store          Liquor Store            Restaurant   
15           Coffee Shop    Italian Restaurant        Breakfast Spot   
16           Coffee Shop                   Pub         Grocery Store   

   5th Most Common Venue          6th Most Common Venue  \
0                   Park                        Brewery   
1          Grocery Store                            Gym   
2     Italian Restaurant  Vegetarian / Vegan Restaurant   
3     Italian Restaurant                    Pizza Place   
4            Pizza Place                    Golf Course   
5            Pizza Place                  Grocery Store   
10              Pharmacy                            Gym   
11                  Café                     Restaurant   
12     Convenience Store                    Yoga Studio   
13          Skating Rink                     Campground   
14                   Gym                    Pizza Place   
15                  Café                           Park   
16              Pharmacy                 Breakfast Spot   

            7th Most Common Venue  8th Most Common Venue  \
0                       Gastropub            Pizza Place   
1                      Restaurant           Liquor Store   
2                     Pizza Place          Grocery Store   
3                    Liquor Store          Grocery Store   
4                             Gym               Pharmacy   
5                             Gym         Sandwich Place   
10                    Pizza Place             Restaurant   
11  Vegetarian / Vegan Restaurant          Grocery Store   
12                           Café         Ice Cream Shop   
13             Chinese Restaurant  Portuguese Restaurant   
14                          Diner     Italian Restaurant   
15  Vegetarian / Vegan Restaurant            Pizza Place   
16              Electronics Store                   Bank   

            9th Most Common Venue     10th Most Common Venue  
0   Vegetarian / Vegan Restaurant         Italian Restaurant  
1                            Bank             Sandwich Place  
2                      Restaurant          Indian Restaurant  
3   

In [65]:
#Blue Cluster
London_merged.loc[London_merged['Cluster Labels'] == 2, London_merged.columns[[2] + list(range(6, London_merged.shape[1]))]]

Neighbourhood 1st Most Common Venue 2nd Most Common Venue  \
9    East Tempo                 Hotel    Italian Restaurant   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
9           Golf Course        Breakfast Spot                 Diner   

  6th Most Common Venue 7th Most Common Venue        8th Most Common Venue  \
9        Discount Store               Dog Run  Eastern European Restaurant   

  9th Most Common Venue 10th Most Common Venue  
9          Dessert Shop      Electronics Store

In [66]:
#Turquoise Cluster
London_merged.loc[London_merged['Cluster Labels'] == 3, London_merged.columns[[2] + list(range(6, London_merged.shape[1]))]]

Neighbourhood 1st Most Common Venue  \
6  Talbot,Lambeth,West Tempo,South Sharon Creek              Pharmacy   
8   Jackson,Old Victoria,Bradley,North Highbury           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
6                   Gym         Grocery Store                   Pub   
8           Supermarket           Gas Station              Pharmacy   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
6           Coffee Shop        Farmers Market                  Bank   
8         Grocery Store          Intersection           Flower Shop   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
6           Gas Station        Sandwich Place             Beer Store  
8  Fast Food Restaurant         Movie Theater                Dog Run

### Conclusion